Source/lighting.cpp

```diff
@@ -601,126 +614,67 @@ void SavePreLighting()
 	memcpy(dPreLight, dLight, sizeof(dPreLight));
 }
 
-void InitVision()
-{
-	VisionCount = 0;
-	dovision = false;
-	VisionId = 1;
-
-	for (int i = 0; i < TransVal; i++) {
-		TransList[i] = false;
-	}
-}
-
-int AddVision(Point position, int r, bool mine)
+void AddVision(Point position, int r, int id)
 {
-	if (VisionCount >= MAXVISION)
-		return -1;
-
-	auto &vision = VisionList[VisionCount];
+	auto &vision = VisionList[id];
 	vision.position.tile = position;
-	vision._lradius = r;
-	vision._lid = VisionId;
-	vision._ldel = false;
-	vision._lunflag = false;
-	vision._lflags = mine;
-
-	VisionId++;
-	VisionCount++;
-	dovision = true;
+	vision.radius = r;
+	vision.isInvalid = false;
+	vision.hasChanged = false;
 
-	return vision._lid;
+	UpdateVision = true;
 }
 
 void ChangeVisionRadius(int id, int r)
 {
-	for (int i = 0; i < VisionCount; i++) {
-		auto &vision = VisionList[i];
-		if (vision._lid != id)
-			continue;
-
-		vision._lunflag = true;
-		vision.position.old = vision.position.tile;
-		vision.oldRadius = vision._lradius;
-		vision._lradius = r;
-		dovision = true;
-	}
+	auto &vision = VisionList[id];
+	vision.hasChanged = true;
+	vision.position.old = vision.position.tile;
+	vision.oldRadius = vision.radius;
+	vision.radius = r;
+	UpdateVision = true;
 }
 
 void ChangeVisionXY(int id, Point position)
 {
-	for (int i = 0; i < VisionCount; i++) {
-		auto &vision = VisionList[i];
-		if (vision._lid != id)
-			continue;
-
-		vision._lunflag = true;
-		vision.position.old = vision.position.tile;
-		vision.oldRadius = vision._lradius;
-		vision.position.tile = position;
-		dovision = true;
-	}
+	auto &vision = VisionList[id];
+	vision.hasChanged = true;
+	vision.position.old = vision.position.tile;
+	vision.oldRadius = vision.radius;
+	vision.position.tile = position;
+	UpdateVision = true;
 }
 
 void ProcessVisionList()
 {
-	if (!dovision)
+	if (!UpdateVision)
 		return;
 
-	for (int i = 0; i < VisionCount; i++) {
-		auto &vision = VisionList[i];
-		if (vision._ldel) {
-			DoUnVision(vision.position.tile, vision._lradius);
-		}
-		if (vision._lunflag) {
-			DoUnVision(vision.position.old, vision.oldRadius);
-			vision._lunflag = false;
-		}
-	}
 	for (int i = 0; i < TransVal; i++) {
 		TransList[i] = false;
 	}
-	for (int i = 0; i < VisionCount; i++) {
-		auto &vision = VisionList[i];
-		if (vision._ldel)
-			continue;
 
-		MapExplorationType doautomap = MAP_EXP_SELF;
-		if (!vision._lflags) {
-			doautomap = MAP_EXP_OTHERS;
-			for (const Player &player : Players) {
-				// Find player for this vision
-				if (!player.plractive || !player.isOnActiveLevel() || player._pvid != vision._lid)
-					continue;
-				// Check that player allows automap sharing
-				if (!player.friendlyMode)
-					doautomap = MAP_EXP_NONE;
-				break;
-			}
+	for (const Player &player : Players) {
+		Light &vision = VisionList[player.getId()];
+		if (!player.plractive || !player.isOnActiveLevel()) {
+			DoUnVision(vision.position.tile, vision.radius);

```

Comment:
Yes, I was thinking it wouldn't be to slow as most of the time the players will be on the same level and avoiding it would require adding some more stuff for tracking it. I could maybe add a bool[4] for tracking if the vision has been disabled already. Again kinda skipped a bit on this since i hope to fully remove it :)

Reply:
I thought about reusing `hasChanged`.
But I'm also fine with letting it that way if we remove it soon.